### ISSUE TITLE
SegFault for some Requests

### ISSUE THREAD

**Issue by adrienjuguet** [Project-OSRM/osrm-backend#2349]
Hello,
I used your work since a week.
I created a server but for some combinations of coordinates the server stops with a segmentation fault.
I try with 2 different profiles (car and foot) and this problem appears just with car profile.

Example of a bad query :
http://127.0.0.1:5000/trip/v1/driving/4.7886357,45.756145;4.8008547,45.74967 

Example of a good query :
http://127.0.0.1:5000/trip/v1/driving/4.7878857,45.756145;4.8007174,45.749783

**Comment by adrienjuguet** [comment]
I noticed that this problem is with trip service and not with route.

**Comment by MoKob** [comment]
Turns out this is due to https://github.com/Project-OSRM/osrm-backend/blob/master/src/engine/plugins/trip.cpp#L140. The route fails if no-turn is allowed at one of the vias in between.
The connected roads to one of the two locations seem to be tagged wrong. Still the implementation should be able to find the u-turn at this point: http://www.openstreetmap.org/#map=18/45.75589/4.78896 .
Investigating further

**Comment by MoKob** [comment]
@adrienjuguet so technically this is a data issue. OSRM in general is designed to not take u-turns, unless we are at a cul de sac. The one way-point in question is located at a drive through (not considered routable) which is tagged incorrectly and at a not routable oneway street (not connected to the road network). Nevertheless, the debug assertion shouldn't trigger.

**Comment by MoKob** [comment]
The problem here is how we discover location in which we allow u-turns.
Just as https://github.com/Project-OSRM/osrm-backend/blob/master/features/testbot/via.feature#L134-L149 shows, this is currently only the case if we have a oneway segment that is a dead end. 

Our current fix is more than hacked and only works for a single segment.

```
    Scenario: Via point at a dead end with oneway
        Given the node map
            | a | b | c |
            |   | d |   |
            |   | e |   |
            |   | f |   |

        And the ways
            | nodes | oneway |
            | abc   |  no    |
            | bd    |  no    |
            | de    |  yes   |
            | ef    |  yes   |

        When I route I should get
            | waypoints | route                |
            | a,d,c     | abc,bd,bd,bd,abc,abc |
            | c,d,a     | abc,bd,bd,bd,abc,abc |
```
Results in 
```
    When I route I should get                     # features/step_definitions/routing.js:4
      | waypoints | route                |
      | a,d,c     | abc,bd,bd,bd,abc,abc |
      | c,d,a     | abc,bd,bd,bd,abc,abc |
      Tables were not identical:
          waypoints |     route
      | (-) a,d,c | (-) abc,bd,bd,bd,abc,abc | 
      | (+) a,d,c | (+)  | 
      | (-) c,d,a | (-) abc,bd,bd,bd,abc,abc | 
      | (+) c,d,a | (+)  |
```

I have prepared a symptomatic fix for the situation triggered here in https://github.com/Project-OSRM/osrm-backend/pull/2350, for a full fix, we would require knowledge on small components.
We need to allow u-turns on all edges entering small components.

**Comment by TheMarex** [comment]
@MoKob can we insert a uturn for every intersection where there are only outgoing edges? That should mitigate the problem with a minimal amount of additional edges.

**Comment by MoKob** [comment]
@TheMarex What do you mean by only outgoing. Nodes that can only be left are not really a problem that is related here, because they form a small component themselves. Problematic are locations that can only be entered. In addition, we also have locations but the only way to exit them is heading into a small component. On all of these we would have to add a u-turn to make sure assumptions about paths in large components to work.

**Comment by MoKob** [comment]
Fixed with #2350
